Relabel BTRFS Content on container Creation

rhatdan opened the pull request:
This change will allow us to run SELinux in a container with
BTRFS back end.  We continue to work on fixing the kernel/BTRFS
but this change will allow SELinux Security separation on BTRFS.

It basically relabels the content on container creation.

Just relabling -init directory in BTRFS use case. Everything looks like it
works. I don't believe tar/achive stores the SELinux labels, so we are good
as far as docker commit.

Tested Speed on startup with BTRFS on top of loopback directory. BTRFS
not on loopback should get even better perfomance on startup time.  The
more inodes inside of the container image will increase the relabel time.

This patch will give people who care more about security the option of
runnin BTRFS with SELinux.  Those who don't want to take the slow down
can disable SELinux either in individual containers or for all containers
by continuing to disable SELinux in the daemon.

Without relabel:

> time docker run --security-opt label:disable fedora echo test
> test

real    0m0.918s
user    0m0.009s
sys    0m0.026s

With Relabel

test

real    0m1.942s
user    0m0.007s
sys    0m0.030s

Signed-off-by: Dan Walsh dwalsh@redhat.com

rhatdan commented:
@tiborvass @crosbymichael @tianon @jfrazelle PTAL

LK4D4 commented:
makes sense for me
ping @docker/core-maintainers

calavera commented:
Moved to code-review, unfortunately it needs a rebase.

Change LGTM.

rhatdan commented:
@calavera Rebased.

calavera commented:
it doesn't look like this compiles any more. Check https://jenkins.dockerproject.org/job/Docker-PRs/16830/console

rhatdan commented:
How do I fix the golint problems with Windows?

calavera commented:
I've opened #16666 to fix those issues.

rhatdan commented:
Don't see how the Janky failures are related to my patch?

calavera commented:
@rhatdan janky runs go-lint in every changed file you commit. For some reason, it looks like `graph/windows/windows.go` never went through go-lint before and it failed when your changes trigger the lint. It should be fixed if you rebased master into this branch.

rhatdan commented:
rebased @calavera Any progress on this?

jfrazelle commented:
```
---> Making bundle: binary (in bundles/1.9.0-dev/binary)
Building: bundles/1.9.0-dev/binary/docker-1.9.0-dev
# github.com/docker/docker/daemon/graphdriver
daemon/graphdriver/plugin.go:32: cannot use proxy (type *graphDriverProxy) as type Driver in return argument:
	*graphDriverProxy does not implement Driver (wrong type for Create method)
		have Create(string, string) error
		want Create(string, string, string) error
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-experimental] $ /bin/sh -xe /tmp/hudson466336133504869178.sh
+ set +x
```

rhatdan commented:
@jfrazelle fixed

cpuguy83 commented:
./docker_cli_external_graphdriver_unix_test.go:127: not enough arguments in call to driver.Create

cpuguy83 commented:
daemon\graphdriver\windows\windows.go:133: d.resolveId undefined (type *Driver has no field or method resolveId)

thaJeztah commented:
ping @rhatdan can you rebase and address the errors?

rhatdan committed:
Relabel BTRFS Content on container Creation

This change will allow us to run SELinux in a container with
BTRFS back end.  We continue to work on fixing the kernel/BTRFS
but this change will allow SELinux Security separation on BTRFS.

It basically relabels the content on container creation.

Just relabling -init directory in BTRFS use case. Everything looks like it
works. I don't believe tar/achive stores the SELinux labels, so we are good
as far as docker commit.

Tested Speed on startup with BTRFS on top of loopback directory. BTRFS
not on loopback should get even better perfomance on startup time.  The
more inodes inside of the container image will increase the relabel time.

This patch will give people who care more about security the option of
runnin BTRFS with SELinux.  Those who don't want to take the slow down
can disable SELinux either in individual containers or for all containers
by continuing to disable SELinux in the daemon.

Without relabel:

> time docker run --security-opt label:disable fedora echo test
test

real    0m0.918s
user    0m0.009s
sys    0m0.026s

With Relabel

test

real    0m1.942s
user    0m0.007s
sys    0m0.030s

Signed-off-by: Dan Walsh <dwalsh@redhat.com>

Signed-off-by: Dan Walsh <dwalsh@redhat.com>

cpuguy83 commented:
A few more errors.

rhatdan commented:
@cpuguy83 In order to get this patch to work.  I needed to reverse the order of

daemon.setHostConfig before daemon.createRootfs which is causing the problems. I am having someone on my team look into why this is causing the tests to be broken, since I am traveling.

We need the order reversed, because with this patch

daemon.createRootfs  calls graphdriver create which needs the SELinux labels initialized.  These are initialized in daemon.setHostConfig, if they are not reversed the labels do not get generated until too late.

thaJeztah commented:
@rhatdan any update on that? going through PRs to see if we can get some movement in the older ones :smile:

rhatdan commented:
@thaJeztah Sorry been travelling, will start to look into this today.  Problem is the reversing of the 

```
@@ -87,10 +87,10 @@ func (daemon *Daemon) create(params *ContainerCreateConfig) (retC *Container, re
 	if err := daemon.Register(container); err != nil {
 		return nil, err
 	}
-	if err := daemon.createRootfs(container); err != nil {
+	if err := daemon.setHostConfig(container, params.HostConfig); err != nil {
 		return nil, err
 	}
-	if err := daemon.setHostConfig(container, params.HostConfig); err != nil {
+	if err := daemon.createRootfs(container); err != nil {
 		return nil, err
 	}
 	defer func() {
```
Anyone have any idea why this is causing problem.

rhatdan commented:
I moved the security check out of setHostConfig so I can call it before createRootFS.  Hopefully this fixes the test issues.

calavera commented:
LGTM

LK4D4 commented:
LGTM

dmcgowan commented:
Why were the labels added to the `Create` method instead of putting it on `Get`? The interface to `Create` was updated to accommodate this but the reason is not clear.

rhatdan commented:
You need the label at create time in order to recursively set the labels.  Get is too late.  We need the same thing to get labels for OverlayFS if the kernel ever gets fixed.

dmcgowan commented:
The label is only added to one `Create` call, and the next line has a call to `Get`. https://github.com/docker/docker/pull/16452/files#diff-1a1f3e7ad9b1d7584e2d3e7d0c4c3db9R995

Considering this label is done at the very end of `Create`, I don't see anything based on the code changes that wouldn't allow it to be done at the beginning of `Get`.

dmcgowan commented:
I understand the difference, I am asking because this effects #17924 and we have to rebase and make interface updates as a result of this change. From my perspective, this is a btrfs specific change (possibly something similar for overlay in the future) which has now updated the interface to add mount labels on create, where previously create was not intended to mount. If relabeling must happen before first mount on btrfs then it is preferable to limit the change if that is possible. Except for on `daemon.Mount`, `graphdriver.Get` is normally called without a mount label, meaning this would only apply to start. From your previous comment, you are saying the main reason this must happen on create is that calling relabel for every start is either non-performant or is the incorrect behavior.

Somewhat separate question, why does this only apply to the init layer for the container and not any of the base layers or container writable layer?

rhatdan commented:
Yes relabeling is going to be time consuming.  We time it out at about 1-2 seconds on a base image, but it could grow depending on the number of inodes in the image.  Since we only need to do this on create, no need to add this penalty to all Starts.

By only applying it to the init layer. we end up being able to continue to share the image. IE Does not cause a COW change for every container.  @rhvgoyal correct?

rhvgoyal commented:
@rhatdan right. If we apply label on container rootfs (and not init layer), then `docker commit` will make all the image files part of the COW layer and that's not what we want.
